volume/modules/local/grpquota.go

```diff
@@ -0,0 +1,313 @@
+		return mountPoint, fmt.Errorf("mountPoint not found: %s", dir)
+	}
+	if !hasQuota {
+		_, _, _, err := exec.Run(0, "mount", "-o", "remount,grpquota", mountPoint)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	vfsVersion, quotaFilename, err := quota.getVFSVersionAndQuotaFile(devID)
+	if err != nil {
+		return "", err
+	}
+
+	filename := mountPoint + "/" + quotaFilename
+	if _, err := os.Stat(filename); err != nil {
+		os.Remove(mountPoint + "/aquota.user")
+
+		header := []byte{0x27, 0x19, 0xc0, 0xd9, 0x00, 0x00, 0x00, 0x00, 0x80, 0x3a, 0x09, 0x00, 0x80,
+			0x3a, 0x09, 0x00, 0x00, 0x00, 0x00, 0x00, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x05, 0x00, 0x00, 0x00}
+		if vfsVersion == "vfsv1" {
+			header[4] = 0x01
+		}
+
+		if writeErr := ioutil.WriteFile(filename, header, 644); writeErr != nil {
+			logrus.Errorf("write file error. %s, %s, %s", filename, vfsVersion, writeErr)
+			return mountPoint, writeErr
+		}
+		if _, _, _, err := exec.Run(0, "setquota", "-g", "-t", "43200", "43200", mountPoint); err != nil {
+			os.Remove(filename)
+			return mountPoint, err
+		}
+		if err := quota.setUserQuota(0, 0, mountPoint); err != nil {
+			os.Remove(filename)
+			return mountPoint, err
+		}
+	}
+
+	// on
+	exit, stdout, stderr, err := exec.Run(0, "quotaon", "-pg", mountPoint)
+	if err != nil && exit != 1 {
+		logrus.Errorf("quotaon failed, exit: %d, stdout: %s, stderr: %s, err: %v", exit, stdout, stderr, err)
+		return "", fmt.Errorf("stderr: %s, err: %v", stderr, err)
+	}
+	if strings.Contains(stdout, " is on") {
+		mountPoints[devID] = mountPoint
+		return mountPoint, nil
+	}
+	if _, _, _, err = exec.Run(0, "quotaon", mountPoint); err != nil {
+		mountPoint = ""
+	}
+
+	mountPoints[devID] = mountPoint
+	return mountPoint, err
+}
+
+// SetSubtree is used to set quota id for directory,
+// setfattr -n system.subtree -v $QUOTAID
+func (quota *GrpQuota) SetSubtree(dir string, qid uint32) (uint32, error) {
+	if !UseQuota {
+		return 0, nil
+	}
+
+	id := qid
+	var err error
+	if id == 0 {
+		id = quota.GetFileAttr(dir)
+		if id > 0 {
+			return id, nil
+		}
+		id, err = quota.GetNextQuatoID()
+	}
+
+	if err != nil {
+		return 0, err
+	}
+	strid := strconv.FormatUint(uint64(id), 10)
+	_, _, _, err = exec.Run(0, "setfattr", "-n", "system.subtree", "-v", strid, dir)
+	return id, err
+}
+
+// SetDiskQuota is used to set quota for directory.
+func (quota *GrpQuota) SetDiskQuota(dir string, size string, quotaID int) error {
+	if !UseQuota {
+		return nil
+	}
+
+	mountPoint, err := quota.StartQuotaDriver(dir)
+	if err != nil {
+		return err
+	}
+	if len(mountPoint) == 0 {
+		return fmt.Errorf("mountpoint not found: %s", dir)
+	}
+
+	id, err := quota.SetSubtree(dir, uint32(quotaID))
+	if id == 0 {
+		return fmt.Errorf("subtree not found: %s %v", dir, err)
+	}
+
+	limit, err := bytefmt.ToKilobytes(size)
+	if err != nil {
+		return err
+	}
+
+	return quota.setUserQuota(id, limit, mountPoint)
+}
+
+// CheckMountpoint is used to check mount point.
```

Comment:
Could you add one example valid line in /proc/mounts in comment please?
like:
```
// CheckMountpoint is used to check mount point which is in file /proc/mounts.
// one valid line must has the following format:
// cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb 0 0
```
You can also tell others more about the exact meanings of each column. 😄

Reply:
ok